.azure-pipelines/test_plan.py

```diff
@@ -182,6 +183,29 @@ def cmd(self, cmds):
         return stdout, stderr, return_code
 
     def get_token(self):
+
+        # Success of "az account get-access-token" depends on "az login". However, the "az login" session may expire
+        # after 24 hours. So we re-login every 12 hours to ensure success of "az account get-access-token".
+        if datetime.now() - self.init_time > timedelta(hours=12):
+            cmd = "az login --service-principal -u {} --tenant {} --allow-no-subscriptions --federated-token {}"\
```

Comment:
How about merging the logic with the generate token, so that we don't need to duplicate the time check logic?
like:
```
token_is_valid = \
            self._token_expires_on is not None and \
            (self._token_expires_on - datetime.now()) > timedelta(hours=TOKEN_EXPIRE_HOURS)
        if self._token is not None and token_is_valid:
            return self._token
        #do login
        cmd = 'az login.....'
        #do get token
        cmd = 'az account get-access-token --resource {}'.format(self.client_id)
```

Reply:
The token expiration and `az login` session expiration are different. I don't think they are duplicated logic.

Reply:
Nit: I guess a small bit of refactoring might help reuse the command execution bit. Not much of saving but 3-4 lines of duplicate code - 

```
def az_run(self, cmd):
    stdout, stderr, retcode = self.cmd(cmd.split())
    if retcode != 0:
        raise Exception(f'Command {cmd} execution failed with error code {rc}')
    return stdout, stderr, retcode
```

In the caller place -

```
while attempt < MAX_GET_TOKEN_RETRY_TIMES:
    try:
         stdout, stderr, retcode = self.az_run(cmd)
         self.last_login_time = datetime.now()
     except Exception as e:
          attempt += 1
```

Reply:
Good suggestion